src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ParameterCapturingService.cs

```diff
@@ -1,81 +1,270 @@
+        private readonly ParameterCapturingEventSource _eventSource = new();
         private readonly ILogger? _logger;
 
+        private Channel<StartCapturingParametersPayload>? _requests;
+        private Channel<bool>? _stopRequests;
+
         public ParameterCapturingService(IServiceProvider services)
         {
-            _logger = services.GetService<ILogger<ParameterCapturingService>>();
-            if (_logger == null)
+            try
+            {
+                ArgumentNullException.ThrowIfNull(SharedInternals.MessageDispatcher);
+
+                // Register the command callbacks (if possible) first so that dotnet-monitor
+                // can be notified of any initialization errors when it tries to invoke the commands.
+                SharedInternals.MessageDispatcher.RegisterCallback<StartCapturingParametersPayload>(
+                    IpcCommand.StartCapturingParameters,
+                    OnStartMessage);
+
+                SharedInternals.MessageDispatcher.RegisterCallback<EmptyPayload>(
+                    IpcCommand.StopCapturingParameters,
+                    OnStopMessage);
+
+                _logger = services.GetService<ILogger<ParameterCapturingService>>();
+                if (_logger == null)
+                {
+                    throw new NotSupportedException(ParameterCapturingStrings.FeatureUnsupported_NoLogger);
+                }
+
+                //
+                // Request processing overview:
+                //
+                // - Incoming Requests
+                //   - dotnet-monitor will properly rate limit requests so there will never be 2 start requests at any given time.
+                //   - Our monitor message processing (start/stop commands) happens serially.
+                //   - Our monitor message processing simply ACKs a command if there's nothing immediately wrong and queues the
+                //     work to happen on our background service.
+                // - Request Handling
+                //  - Happens on our background service.
+                //  - The background service may stop at any given point if an unrecoverable error occurs. In this scenario,
+                //    dotnet-monitor is notified of the error, and all future requests will short-circuit and notify
+                //    dotnet-monitor that the parameter capturing service is unavailable.
+                //
+                _requests = Channel.CreateBounded<StartCapturingParametersPayload>(new BoundedChannelOptions(capacity: 1)
+                {
+                    FullMode = BoundedChannelFullMode.DropWrite
+                });
+                _stopRequests = Channel.CreateBounded<bool>(new BoundedChannelOptions(capacity: 1)
+                {
+                    FullMode = BoundedChannelFullMode.DropWrite
+                });
+
+                _probeManager = new FunctionProbesManager(new LogEmittingProbes(_logger));
+            }
+            catch (Exception ex)
```

Comment:
I understand that you want to make sure that no exception bubbles out of this class so the user's app isn't affected, but the way this is done is causing all of the fields to be nullable (even though the compilation is forcing nullable attribution), which causes the rest of the code to constantly either null check or tell the compiler "no, trust me, it's not null". Let's think of a way to make everything non-nullable to avoid this.